adding keep in memory

Mustapha-AJEGHRIR opened the pull request:
Fixing #514 .

Hello @mariosasko 👋, I have implemented what you have recommanded to fix the keep in memory problem for shuffle on the issue #514 .

mariosasko commented on src/datasets/arrow_dataset.py:
```diff
@@ -3585,7 +3588,7 @@ def shuffle(
 
         # Check if we've already cached this computation (indexed by a hash)
-        if self.cache_files:
+        if self.cache_files and not keep_in_memory:
```
This line should stay the same:

```suggestion
        if self.cache_files:
```

Instead, [this](https://github.com/huggingface/datasets/blob/7d51531c58127d44565d6d1a6a7aa75e55d2b036/src/datasets/arrow_dataset.py#L3603) line should be replaced with:
```python
indices_cache_file_name=indices_cache_file_name if not keep_in_memory else None
```

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/datasets/pr_5082). All of your documentation changes will be reflected on that endpoint.

mariosasko reviewed (commented):
Thanks!

We also need a test with `shuffle(keep_in_memory=True)` for this fix. Let me know if you need help with that.

Mustapha-AJEGHRIR commented:
Hi @mariosasko , I have added a test for the `keep_in_memory` version. I have also removed the `Compatible with temp_seed` part in the scope of `dset_shuffled`, please verify if that makes sense.

mariosasko reviewed (approved):
Looks all good now!

lhoestq reviewed (approved):
Nice !
